File: src/stores/simulation.ts
import { create } from "zustand";
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

// --- Types ---

export interface SolverAvailability {
  ngspice: boolean;
  ngspice_path: string | null;
  openems: boolean;
  openems_path: string | null;
  elmer: boolean;
  elmer_path: string | null;
}

export interface WaveformData {
  name: string;
  unit: string;
  real: number[];
  imag: number[] | null;
}

export interface SimulationResult {
  sim_id: string;
  analysis_type: string;
  vectors: Record<string, WaveformData>;
  elapsed_ms: number;
  netlist: string;
}

export interface SimProgress {
  sim_id: string;
  percent: number;
  message: string;
}

export type AnalysisType = "DcOp" | "DcSweep" | "Ac" | "Transient" | "Noise";

export interface AnalysisConfig {
  analysis_type: AnalysisType;
  params: Record<string, string>;
}

// --- Store ---

interface SimulationState {
  // Solver availability
  solvers: SolverAvailability;
  solversDetected: boolean;

  // Active simulation
  isRunning: boolean;
  progress: number;
  progressMessage: string;

  // Results
  results: Record<string, SimulationResult>;
  activeResultId: string | null;

  // Analysis config
  analysisType: AnalysisType;
  analysisParams: Record<string, string>;

  // Waveform viewer
  selectedVectors: string[];
  cursorX: number | null;

  // Netlist preview
  netlistPreview: string | null;

  // Actions
  detectSolvers: () => Promise<void>;
  setAnalysisType: (type_: AnalysisType) => void;
  setAnalysisParam: (key: string, value: string) => void;
  runSimulation: (data: unknown) => Promise<void>;
  previewNetlist: (data: unknown) => Promise<void>;
  selectResult: (simId: string) => void;
  toggleVector: (name: string) => void;
  setCursorX: (x: number | null) => void;
  clearResults: () => void;
}

// Module-level listener cleanup
let progressUnlisten: UnlistenFn | null = null;

export const useSimulationStore = create<SimulationState>((set, get) => ({
  solvers: {
    ngspice: false, ngspice_path: null,
    openems: false, openems_path: null,
    elmer: false, elmer_path: null,
  },
  solversDetected: false,
  isRunning: false,
  progress: 0,
  progressMessage: "",
  results: {},
  activeResultId: null,
  analysisType: "DcOp",
  analysisParams: {},
  selectedVectors: [],
  cursorX: null,
  netlistPreview: null,

  detectSolvers: async () => {
    try {
      const solvers = await invoke<SolverAvailability>("detect_solvers");
      set({ solvers, solversDetected: true });
    } catch (e) {
      console.error("Failed to detect solvers:", e);
    }
  },

  setAnalysisType: (type_) => set({ analysisType: type_, analysisParams: {} }),
  setAnalysisParam: (key, value) =>
    set((s) => ({ analysisParams: { ...s.analysisParams, [key]: value } })),

  runSimulation: async (data) => {
    const { analysisType, analysisParams } = get();
    const simId = `sim_${Date.now()}`;

    set({ isRunning: true, progress: 0, progressMessage: "Starting..." });

    // Listen for progress events
    if (progressUnlisten) progressUnlisten();
    progressUnlisten = await listen<SimProgress>("sim:progress", (event) => {
      set({
        progress: event.payload.percent,
        progressMessage: event.payload.message,
      });
    });

    try {
      const result = await invoke<SimulationResult>("run_spice_simulation", {
        simId,
        data,
        config: { analysis_type: analysisType, params: analysisParams },
      });

      set((s) => ({
        results: { ...s.results, [result.sim_id]: result },
        activeResultId: result.sim_id,
        selectedVectors: Object.keys(result.vectors).filter(
          (n) => n.toLowerCase() !== "time" && n.toLowerCase() !== "frequency"
        ),
        isRunning: false,
        progress: 100,
        progressMessage: `Done in ${result.elapsed_ms}ms`,
      }));
    } catch (e) {
      set({
        isRunning: false,
        progress: 0,
        progressMessage: `Error: ${e}`,
      });
    } finally {
      if (progressUnlisten) {
        progressUnlisten();
        progressUnlisten = null;
      }
    }
  },

  previewNetlist: async (data) => {
    const { analysisType, analysisParams } = get();
    try {
      const netlist = await invoke<string>("get_spice_netlist", {
        data,
        config: { analysis_type: analysisType, params: analysisParams },
      });
      set({ netlistPreview: netlist });
    } catch (e) {
      set({ netlistPreview: `Error: ${e}` });
    }
  },

  selectResult: (simId) => {
    const result = get().results[simId];
    if (result) {
      set({
        activeResultId: simId,
        selectedVectors: Object.keys(result.vectors).filter(
          (n) => n.toLowerCase() !== "time" && n.toLowerCase() !== "frequency"
        ),
      });
    }
  },

  toggleVector: (name) =>
    set((s) => ({
      selectedVectors: s.selectedVectors.includes(name)
        ? s.selectedVectors.filter((n) => n !== name)
        : [...s.selectedVectors, name],
    })),

  setCursorX: (x) => set({ cursorX: x }),

  clearResults: () => set({ results: {}, activeResultId: null, selectedVectors: [] }),
}));

File: src-tauri/src/commands/simulation.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Mutex;
use tauri::Emitter;

use crate::engine::ngspice_ffi::{self, NgspiceInstance, NgspiceMessage};
use crate::engine::parser::SchematicSheet;
use crate::engine::spice_netlist::{self, AnalysisConfig, AnalysisType};

// --- Types ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SolverAvailability {
    pub ngspice: bool,
    pub ngspice_path: Option<String>,
    pub openems: bool,
    pub openems_path: Option<String>,
    pub elmer: bool,
    pub elmer_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WaveformData {
    pub name: String,
    pub unit: String,
    pub real: Vec<f64>,
    pub imag: Option<Vec<f64>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimulationResult {
    pub sim_id: String,
    pub analysis_type: String,
    pub vectors: HashMap<String, WaveformData>,
    pub elapsed_ms: u64,
    pub netlist: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimProgress {
    pub sim_id: String,
    pub percent: f64,
    pub message: String,
}

// --- Global ngspice instance ---
// ngspice is single-threaded; we hold one instance behind a Mutex.
static NGSPICE: std::sync::LazyLock<Mutex<Option<NgspiceInstance>>> =
    std::sync::LazyLock::new(|| Mutex::new(None));

fn ensure_ngspice() -> Result<(), String> {
    let mut guard = NGSPICE.lock().map_err(|e| format!("Lock error: {}", e))?;
    if guard.is_some() {
        return Ok(());
    }
    let path = ngspice_ffi::detect_ngspice()
        .ok_or_else(|| "ngspice shared library not found. Install ngspice or add it to PATH.".to_string())?;
    let instance = NgspiceInstance::new(&path)?;
    *guard = Some(instance);
    Ok(())
}

// --- Commands ---

#[tauri::command]
pub fn detect_solvers() -> SolverAvailability {
    let ngspice_path = ngspice_ffi::detect_ngspice();

    // Check for OpenEMS
    let openems_path = which_executable("openEMS");

    // Check for Elmer
    let elmer_path = which_executable("ElmerSolver");

    SolverAvailability {
        ngspice: ngspice_path.is_some(),
        ngspice_path: ngspice_path.map(|p| p.to_string_lossy().into_owned()),
        openems: openems_path.is_some(),
        openems_path: openems_path.map(|p| p.to_string_lossy().into_owned()),
        elmer: elmer_path.is_some(),
        elmer_path: elmer_path.map(|p| p.to_string_lossy().into_owned()),
    }
}

#[tauri::command]
pub fn get_spice_netlist(
    data: SchematicSheet,
    config: AnalysisConfig,
) -> Result<String, String> {
    spice_netlist::generate_spice_netlist(&data, &config)
}

#[tauri::command]
pub async fn run_spice_simulation(
    app: tauri::AppHandle,
    sim_id: String,
    data: SchematicSheet,
    config: AnalysisConfig,
) -> Result<SimulationResult, String> {
    let start = std::time::Instant::now();

    // Generate netlist
    let netlist = spice_netlist::generate_spice_netlist(&data, &config)?;

    // Emit progress
    let _ = app.emit(
        "sim:progress",
        SimProgress {
            sim_id: sim_id.clone(),
            percent: 10.0,
            message: "Netlist generated".to_string(),
        },
    );

    // Initialize ngspice if needed
    ensure_ngspice()?;

    let _ = app.emit(
        "sim:progress",
        SimProgress {
            sim_id: sim_id.clone(),
            percent: 20.0,
            message: "ngspice initialized".to_string(),
        },
    );

    // Run simulation in blocking task (ngspice is not async)
    let netlist_clone = netlist.clone();
    let sim_id_clone = sim_id.clone();
    let app_clone = app.clone();
    let analysis_type_str = format!("{:?}", config.analysis_type);

    let result = tokio::task::spawn_blocking(move || {
        let guard = NGSPICE.lock().map_err(|e| format!("Lock error: {}", e))?;
        let ng = guard
            .as_ref()
            .ok_or_else(|| "ngspice not initialized".to_string())?;

        // Reset previous state
        let _ = ng.reset();

        // Load circuit from netlist lines
        let lines: Vec<&str> = netlist_clone.lines().collect();
        ng.load_circuit(&lines)?;

        let _ = app_clone.emit(
            "sim:progress",
            SimProgress {
                sim_id: sim_id_clone.clone(),
                percent: 40.0,
                message: "Circuit loaded, running simulation...".to_string(),
            },
        );

        // Run the simulation
        ng.command("run")?;

        // Drain callback messages (progress updates)
        while let Ok(msg) = ng.rx.try_recv() {
            match msg {
                NgspiceMessage::Status(s) => {
                    // Parse percent from status string like "--ready"
                    let _ = app_clone.emit(
                        "sim:progress",
                        SimProgress {
                            sim_id: sim_id_clone.clone(),
                            percent: 60.0,
                            message: s,
                        },
                    );
                }
                NgspiceMessage::Output(s) => {
                    // Log output for debugging
                    let _ = app_clone.emit(
                        "sim:progress",
                        SimProgress {
                            sim_id: sim_id_clone.clone(),
                            percent: 60.0,
                            message: s,
                        },
                    );
                }
                _ => {}
            }
        }

        let _ = app_clone.emit(
            "sim:progress",
            SimProgress {
                sim_id: sim_id_clone.clone(),
                percent: 80.0,
                message: "Simulation complete, extracting results...".to_string(),
            },
        );

        // Get current plot and all vectors
        let plot_name = ng
            .current_plot()
            .ok_or_else(|| "No current plot after simulation".to_string())?;
        let vec_names = ng.all_vectors(&plot_name);

        let mut vectors: HashMap<String, WaveformData> = HashMap::new();

        for name in &vec_names {
            let full_name = format!("{}.{}", plot_name, name);
            match ng.get_vector(&full_name) {
                Ok((real, imag)) => {
                    let unit = guess_unit(name);
                    vectors.insert(
                        name.clone(),
                        WaveformData {
                            name: name.clone(),
                            unit,
                            real,
                            imag,
                        },
                    );
                }
                Err(_) => {
                    // Try without plot prefix
                    if let Ok((real, imag)) = ng.get_vector(name) {
                        let unit = guess_unit(name);
                        vectors.insert(
                            name.clone(),
                            WaveformData {
                                name: name.clone(),
                                unit,
                                real,
                                imag,
                            },
                        );
                    }
                }
            }
        }

        Ok::<HashMap<String, WaveformData>, String>(vectors)
    })
    .await
    .map_err(|e| format!("Simulation task failed: {}", e))??;

    let elapsed = start.elapsed().as_millis() as u64;

    let sim_result = SimulationResult {
        sim_id: sim_id.clone(),
        analysis_type: analysis_type_str,
        vectors: result,
        elapsed_ms: elapsed,
        netlist,
    };

    let _ = app.emit(
        "sim:progress",
        SimProgress {
            sim_id,
            percent: 100.0,
            message: format!("Done in {}ms", elapsed),
        },
    );

    Ok(sim_result)
}

// --- Helpers ---

fn guess_unit(vec_name: &str) -> String {
    let name_lower = vec_name.to_lowercase();
    if name_lower.starts_with("v(") || name_lower.starts_with("v_") {
        "V".to_string()
    } else if name_lower.starts_with("i(") || name_lower.starts_with("i_") {
        "A".to_string()
    } else if name_lower == "time" {
        "s".to_string()
    } else if name_lower == "frequency" {
        "Hz".to_string()
    } else {
        String::new()
    }
}

/// Search PATH for an executable.
fn which_executable(name: &str) -> Option<std::path::PathBuf> {
    if let Ok(path_var) = std::env::var("PATH") {
        let sep = if cfg!(windows) { ';' } else { ':' };
        for dir in path_var.split(sep) {
            let candidate = std::path::Path::new(dir).join(name);
            if candidate.exists() {
                return Some(candidate);
            }
            // Windows: also check with .exe extension
            #[cfg(windows)]
            {
                let with_ext = candidate.with_extension("exe");
                if with_ext.exists() {
                    return Some(with_ext);
                }
            }
        }
    }
    None
}
